[Go] `unsafe` Varint fast-path writes/reads more bytes than reserved in struct serializer

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fory/issues) and found no similar issues.


### Version

0.17.0

### Component(s)

Go

### Minimal reproduce step

The following self-contained Go test demonstrates the write-path issue using a sentinel-value check. It can be added to go/fory/buffer_test.go:

```go
func TestUnsafePutVarUint32OOBWrite(t *testing.T) {
    // Construct a ByteBuffer backed by a slice where len == writerIndex + 5.
    // This replicates the condition where Reserve(5) is satisfied without expanding
    // the buffer, but UnsafePutVarUint32 then writes 8 physical bytes.

    const sentinelByte = byte(0xAB)
    const totalCap = 16 // backing large enough to observe the overwrite safely
    backing := make([]byte, totalCap, totalCap)

    // Fill bytes [5, totalCap) with sentinel values.
    for i := 5; i < totalCap; i++ {
        backing[i] = sentinelByte
    }

    // Simulate struct.go's fast-path: Reserve(MaxVarintSize=5) on a tight buffer.
    // Expose only 5 bytes of len so Reserve(5) returns immediately.
    buf := NewByteBuffer(backing[:5]) // len=5, cap=16

    // Replicate exactly what struct.go:342 does for a uint32 varint field:
    buf.Reserve(5) // needed(5) <= len(5) → returns immediately, NO growth

    // A uint32 value >= 2^28 encodes to 5 varint bytes → triggers 8-byte bulk write.
    written := buf.UnsafePutVarUint32(0, 1<<28)
    if written != 5 {
        t.Fatalf("expected 5 bytes written, got %d", written)
    }

    // Bytes at indices 5, 6, 7 are past Reserve(5)'s guaranteed region.
    // If they differ from the sentinel, the 8-byte write overflowed the reserved region.
    overwritten := 0
    for i := 5; i < 8; i++ {
        if backing[i] != sentinelByte {
            t.Logf("backing[%d] = 0x%02X (overwritten, was 0x%02X)", i, backing[i], sentinelByte)
            overwritten++
        }
    }
    if overwritten > 0 {
        t.Errorf("UnsafePutVarUint32 wrote %d byte(s) past Reserve(5)'s guaranteed region "+
            "(contract requires Reserve(8) per buffer.go:661)", overwritten)
    }
}

```


### What did you expect to see?

- Expected: test passes — no bytes written past index 4.

### What did you see instead?

- Actual: backing[5], backing[6], backing[7] are overwritten — 3 bytes past the reserved region.

Note on the test design: `cap=16` is used intentionally so the overflow lands within the backing array and the test runs safely without corrupting other heap objects. In production, when a `ByteBuffer` is created from an inbound network slice via `NewByteBuffer(inboundSlice)` (where `len == cap`), those 3 bytes escape the Go allocation and corrupt adjacent heap memory.

### Anything Else?

The Go runtime's struct serialization fast-path in `struct.go` violates the documented contract of `UnsafePutVarUint32` and `UnsafeReadVarUint32` in `buffer.go`. Specifically:

- `UnsafePutVarUint32` documents that the caller must call `Reserve(8)` (because it performs an 8-byte bulk write for 5-byte varints), but `struct.go` only calls `Reserve(MaxVarintSize)`, which is `5` for `uint32/int32` varint fields.
- `UnsafeReadVarUint32` physically reads 8 bytes, but the fast-path guard in `struct.go` only checks `remaining() >= MaxVarintSize` (which can be 5).
This is a correctness bug / contract violation — the mismatch can cause writes past the end of a buffer's allocated memory in any situation where the buffer's capacity is not generously over-sized (e.g. when a `ByteBuffer` is constructed from a pre-existing `[]byte` via `NewByteBuffer`).

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

@chaokunyang 
Here's the failing output - 
```bash
Running tool: /usr/local/go/bin/go test -test.fullpath=true -timeout 30s -run ^TestUnsafePutVarUint32OOBWrite$ github.com/apache/fory/go/fory

--- FAIL: TestUnsafePutVarUint32OOBWrite (0.00s)
    /home/ayush/contributions/apache/fory/go/fory/buffer_test.go:117: backing[5] = 0x00 (overwritten, was 0xAB)
    /home/ayush/contributions/apache/fory/go/fory/buffer_test.go:117: backing[6] = 0x00 (overwritten, was 0xAB)
    /home/ayush/contributions/apache/fory/go/fory/buffer_test.go:117: backing[7] = 0x00 (overwritten, was 0xAB)
    /home/ayush/contributions/apache/fory/go/fory/buffer_test.go:122: UnsafePutVarUint32 wrote 3 byte(s) past Reserve(5)'s guaranteed region (contract requires Reserve(8) per buffer.go:661)
FAIL
FAIL	github.com/apache/fory/go/fory	0.003s
FAIL

```

--- a/go/fory/buffer_test.go
+++ b/go/fory/buffer_test.go
@@ -84,7 +84,7 @@ func checkVarintWrite(t *testing.T, buf *ByteBuffer, value int32) {
 
 // TestUnsafePutVarUint32PhysicalWriteWidth verifies that UnsafePutVarUint32 performs
 // an 8-byte physical write for 5-byte varints and that Reserve(8) (as required by
-// buffer.go:661) keeps those 8 bytes within the backing array.
+// the contract) keeps those 8 bytes within the backing array.
 func TestUnsafePutVarUint32PhysicalWriteWidth(t *testing.T) {
 	const sentinelByte = byte(0xAB)
 	const totalCap = 16

--- a/go/fory/buffer_test.go
+++ b/go/fory/buffer_test.go
@@ -81,3 +81,33 @@ func checkVarintWrite(t *testing.T, buf *ByteBuffer, value int32) {
 	require.Equal(t, buf.ReaderIndex(), buf.WriterIndex())
 	require.Equal(t, value, varInt)
 }
+
+// TestUnsafePutVarUint32PhysicalWriteWidth verifies that UnsafePutVarUint32 performs
+// an 8-byte physical write for 5-byte varints and that Reserve(8) (as required by
+// buffer.go:661) keeps those 8 bytes within the backing array.
+func TestUnsafePutVarUint32PhysicalWriteWidth(t *testing.T) {
+	const sentinelByte = byte(0xAB)
+	const totalCap = 16
+	backing := make([]byte, totalCap, totalCap)
+
+	// Fill [8, totalCap) with sentinels; [0, 8) is the reserved window.
+	for i := 8; i < totalCap; i++ {
+		backing[i] = sentinelByte
+	}
+
+	// Expose 8 bytes of len, matching Reserve(8) contract.
+	buf := NewByteBuffer(backing[:8])
+
+	// Reserve(8) should return immediately as len(data) is already 8.
+	buf.Reserve(8)
+
+	// Encode value >= 2^28 (5 varint bytes) which triggers 8-byte bulk write.
+	written := buf.UnsafePutVarUint32(0, 1<<28)
+	require.Equal(t, 5, written, "expected 5 logical bytes written")
+
+	// Verify bytes [8, totalCap) remain untouched by the 8-byte bulk write.
+	for i := 8; i < totalCap; i++ {
+		require.Equal(t, sentinelByte, backing[i],
+			"byte at index %d is outside the 8-byte reserved window and must not be written", i)
+	}
+}

--- a/go/fory/struct.go
+++ b/go/fory/struct.go
@@ -339,7 +339,8 @@ func (s *structSerializer) WriteData(ctx *WriteContext, value reflect.Value) {
 	// - Reserve max size once, track offset locally, update writerIndex once at end
 	// ==========================================================================
 	if s.fieldGroup.MaxVarintSize > 0 {
-		buf.Reserve(s.fieldGroup.MaxVarintSize)
+		// +8 padding for UnsafePutVarUint32 bulk write (8 bytes physically written for 5-byte varints)
+		buf.Reserve(s.fieldGroup.MaxVarintSize + 8)
 		offset := buf.WriterIndex()
 
 		for _, field := range s.fieldGroup.PrimitiveVarintFields {
@@ -1530,7 +1531,8 @@ func (s *structSerializer) ReadData(ctx *ReadContext, value reflect.Value) {
 	// Note: For tagged int64/uint64, we can't use unsafe reads because they need bounds checking
 	if len(s.fieldGroup.PrimitiveVarintFields) > 0 {
 		err := ctx.Err()
-		if buf.remaining() >= s.fieldGroup.MaxVarintSize {
+		// +8 padding for readVarUint32Fast bulk load (8 bytes physically read regardless of varint length)
+		if buf.remaining() >= s.fieldGroup.MaxVarintSize+8 {
 			for _, field := range s.fieldGroup.PrimitiveVarintFields {
 				fieldPtr := unsafe.Add(ptr, field.Offset)
 				optInfo := optionalInfo{}

--- a/go/fory/struct_test.go
+++ b/go/fory/struct_test.go
@@ -617,3 +617,29 @@ func TestFloat16StructField(t *testing.T) {
 	// Specific value check
 	require.Equal(t, float32(1.5), res.F16.Float32())
 }
+
+// TestVarintFastPathTightBuffer exercises the varint fast-path with a single uint32 field.
+// Serializing a value requiring 5 varint bytes exercises the 8-byte bulk write.
+// Deserializing from a tight buffer (len==cap) exercises the read guard, ensuring
+// the 8-byte bulk load does not read past the end of the backing array.
+func TestVarintFastPathTightBuffer(t *testing.T) {
+	type SingleVarintStruct struct {
+		// compress=true forces the varint fast path.
+		Value uint32 `fory:"compress=true"`
+	}
+
+	f := New(WithXlang(false))
+	require.NoError(t, f.RegisterStruct(SingleVarintStruct{}, 7001))
+
+	// 1<<28 requires 5 varint bytes, forcing the 8-byte bulk write path.
+	obj := SingleVarintStruct{Value: 1 << 28}
+
+	data, err := f.Serialize(&obj)
+	require.NoError(t, err)
+
+	// Deserialize from a tight buffer where len == cap. This ensures the read guard
+	// properly handles bulk loads that would otherwise overrun the slice.
+	var out SingleVarintStruct
+	require.NoError(t, f.Deserialize(data, &out))
+	require.Equal(t, obj.Value, out.Value)
+}

